### ISSUE TITLE
Remove dependency formats from crate metadata

### ISSUE THREAD

**Issue by bjorn3** [rust-lang/rust#96708]
It only really needs to be stored for dylibs and only the dependency format for said dylib and not for any other crate types also used. It can be stored in a separate section for dylibs. For example `.rsdpfmt` (needs to stay within the 8 char limit for COFF section names). Furthermore omitting it from the crate metadata is necessary for implementing https://github.com/bazelbuild/rules_rust/issues/228#issuecomment-1113372651. Not encoding it in the crate metadata may also improve performance a tiny bit although rlib only crates already encode an empty list anyway.

**Comment by davidbarsky** [comment]
I was talking to @jsgf about this issue, but I had two questions:
- Does this issue interact with https://github.com/rust-lang/rust/issues/38913? If so, the work described in 38913 might be less strenuous than expected or this might be more complex than this issue might otherwise imply—can you clarify, if you've got the time?
- If it's a relatively small tweak, does this issue boil down to skipping hashes in https://doc.rust-lang.org/stable/nightly-rustc/rustc_metadata/rmeta/encoder/fn.encode_metadata.html?

To summarize/get everyone on the same page, the reason that @jsgf took the "hollow .rlib" approach (as described in https://github.com/bazelbuild/rules_rust/issues/228#issuecomment-1100788353) was that BuckV2's rules allows for reusing the generated `.rlib` across the equivalent of Buck's `cargo check` and `cargo build` flavors (Jeremy allude to that in the comment, but I didn't pick up on that until he made it explicit to me). 

(On an aside, I've noticed that the approach that Jeremy took incurs about 20-30% more overhead compared to emitting plain metadata, so maybe resolving this issue can hopefully reduce that discrepancy.)

**Comment by bjorn3** [comment]
Currently the crate metadata is shared between all crate types. For dylibs we still need to store the dylib dependency formats. I think the best way is to store it in an additional section of the dylib and then remove it from the main crate metadata. I don't think this interacts with MIR-only rlibs at all.

**Comment by jsgf** [comment]
Just to make sure we're all talking about the same thing, "dependency format" just means [this](https://doc.rust-lang.org/stable/nightly-rustc/rustc_middle/middle/dependency_format/index.html) right?

That's only linkages and crate types, so there's no hashes or any other identifiers of specific crates. So @davidbarsky I don't think this is relevant to any of the things we're talking about.

**Comment by bjorn3** [comment]
> Just to make sure we're all talking about the same thing, "dependency format" just means [this](https://doc.rust-lang.org/stable/nightly-rustc/rustc_middle/middle/dependency_format/index.html) right?

Correct.

**Comment by mati865** [comment]
> For example .rsdpfmt (needs to stay within the 8 char limit for COFF section names).

If this is only for dylibs then longer names for COFF are doable. You can put the actual long name into the string table and then call the section `/<offset>`, where the offset is the offset of the string in the table in decimal ASCII form.

**Comment by bjorn3** [comment]
According to https://github.com/rust-lang/rust/blob/e0d014a3dffbb3f0575cfbeb0f480c5080c4d018/compiler/rustc_codegen_ssa/src/back/metadata.rs#L539-L552 the linker will truncate the section name.

**Comment by mehdiakiki** [comment]
The plan is to move the `dylib_dependency_formats` out of `CrateRoot` and into a format that's dylib i.e. `.rsdpfmt` section and I would read it back when needed in the locator. Claimed it as it is similar to the work done in 156735. Is someone already working on this? and is the `.rsdpfmt` in the metadata wrapper the location that you intended to use?

**Comment by bjorn3** [comment]
> Is someone already working on this?

No. I was planning to work on it later as part of https://github.com/rust-lang/compiler-team/issues/908, but feel free to do it now.

> and is the .rsdpfmt in the metadata wrapper the location that you intended to use?

Yes. I guess we could bikeshed about the best section name, but the best place I think would be the metadata object file. No need to introduce another object file for it. What matters in the end for https://github.com/rust-lang/compiler-team/issues/908 is that the encoding happens during link_binary (probably in `add_local_crate_metadata_objects` or a function called by it) rather than before codegen.
